### DISCUSSION TITLE
Goblin Faction

### DISCUSSION THREAD

**Discussion by evanbowman** [evanbowman/skyland-beta#129] [Ideas]
I intended to start by adding a Sylph faction, but I think I'll add a goblin faction first, as there's less work to do.
Some ideas:
* flimsy cannon: goblin weapon that becomes less powerful as it gets damaged
* Goblin quests: I'd like to eventually replace some of the quest levels with faction-specific ones
* I'm working on replacing some neutral encounters with new events specific to the goblin faction
* Crew icon selection has been updated to be race-specific

**Comment by Doronimmo** [comment]
made some progress on adventure log, mainly useful for goblins since they talk differently

**Reply by Doronimmo** [reply]
currently, there is a seperate txt file for each faction. this allows changes in the log lines, but much else

**Reply by evanbowman** [reply]
It looks good! good idea
```
           Buffer<StringBuffer<20>, 8> args;
            lisp::l_foreach(v->cons().cdr(), [&](lisp::Value* val) {
                lisp::DefaultPrinter p;
                lisp::format(val, p);
                args.push_back(p.data_.c_str());
            });

            const char* fmt_str = str->c_str();
            u32 current_arg = 0;
            while (*fmt_str not_eq '\0') {
                if (*fmt_str == '%') {
                    if (current_arg < args.size()) {
                        result += args[current_arg++];
                    }
                } else {
                    result.push_back(*fmt_str);
                }
                ++fmt_str;
            }
```
This code is repeated a few times, if it's not too much trouble, maybe you could load the line from the file upfront and then we could consolidate the duplicated code a bit
e.g. psuedocode:
``` c++
const char* file = "/strings/adventure_log.txt";
if (APP.faction() == Faction::human) {
    file = "/strings/adventure_log_human.txt";
} else if (..) {
    file = "";
} etc.

auto str = get_line_from_file(file, line);
// the rest of the code.
```

**Reply by Doronimmo** [reply]
65f8583bbe0c0f7361a220222e4d349e7b9d3a9f

**Comment by Doronimmo** [comment]
weapon idea - net-gun: fire at other weaopns, and clog them. weapon trajectory could be a sine (see picture)
![Sprite-0001](https://github.com/user-attachments/assets/0338f2b6-4b92-47dd-a6fa-d2689528a8d9)
it could be interesting for a front firing weapon to be able to hit missiles etc.

**Reply by evanbowman** [reply]
Great idea! I thought about parabolic weapons like that, and had a prototype for one at one point. The issue was that for a parabolic weapon, you sort of have to set both the power and direction of the attack, whereas with the existing weapon selection menu, you can only set the destination target (i.e. the direction of the projectile, but not the magnitude of the attack vector). So I either need to create an additional menu allowing players to choose both the direction and intensity of a parabolic attack, or auto-compute the projectile vector's magnitude based on the layout of the opponent's island and the intended target.

**Reply by evanbowman** [reply]
![ezgif-3e637c20d7e509](https://github.com/user-attachments/assets/67fb74dc-1ae8-40db-8dc8-8bab85613160)
Over the past few days, I've been playing around with this idea more.

**Reply by Doronimmo** [reply]
Looks like some crazy math to calculate the trajectory. **:D**

**Comment by evanbowman** [comment]
While I was hoping not to change too many existing level scripts, I think making more goblin-specific events is a good opportunity to add detail to the game's world. I'm considering moving the dialog for some events to a separate file, and then having folders for each faction with dialog text files.
Some events that need rewriting:

- [x] neutral/1/8.lisp (explorer encounter)
- [x] neutral/1/10.lisp (optional fight with goblin harvester. Maybe you still fight the goblin harvester, but not because you're defending humans)
- [ ] neutral/2/3.lisp (purchase stolen decimator. Maybe replace with some event that lets goblins buy some human exclusive block)
- [ ] neutral/2/6_1.lisp (this one isn't strictly necessary to replace. But talking with some goblins living on the surface would be interesting)
- [ ] The item shops

For the quests... I mainly plan on changing the dialog, setting, island layouts, etc., while preserving the overall gameplay. Keep the skeleton of the level, but make cosmetic changes.
- [x] quest 0 (replace all dialog)
- [x] quest 1 (needs some dialog changes)
- [x] quest 2 (maybe this one can be used with some minor edits, where the humans are slightly apprehensive/distrustful)
- [x] quest 3 (change dialog. Rather than meeting a passenger airship, some other scenario where you pick up goblin crew)
- [x] quest 4
- [x] quest 5 (maybe can still be used, similar to quest 2)
- [x] quest 7 (probably just needs some very minor dialog changes, see comment on quest 2)
- [x] quest 8 (would some human kingdom give goblins an atomic weapon? Perhaps they wouldn't... replace this one)

Misc:
- [ ] Redraw save screen graphics for each faction

Unfortunately, all of the quests, I think, will need to have goblin-specific variants. The idea of merchants entrusting goblins with cargo for delivery, for example, doesn't fit into the game's universe and isn't plausible. This will be quite a bit of work, and a slow process, because I will need to draw new quest artwork. For the sake of stylistic consistency, I think it's important for me to draw this artwork, but it will take some time.

**Reply by evanbowman** [reply]
Furthermore, maybe when playing as goblins, you should have more opportunities to fight human castles. Currently this is not possible in the game, because enemy crew will only display goblin sprites. But eventually, it would be good to perhaps change some of the optional goblin king/queen fights to events where you fight humans. I think it's not strictly necessary yet, because goblins fighting each other is somewhat plausible, and I've set this up thematically by having the goblin castaway hint that you've been banished from the goblin horde for some reason.

**Reply by evanbowman** [reply]
For the goblin harvester encounter: (neutral/1/10.lisp)
Rather than replacing the scenario entirely, I changed it so that when playing as goblins, you can either choose to help the humans by fighting off the other goblins, or refuse to help them and share the spoils with the attacking goblins.

**Comment by Doronimmo** [comment]
Room idea - transporter*: transporter but 2x2 (can transport 2 goblins at once?) and lights the transported room on fire.

**Reply by evanbowman** [reply]
![boarding-capsule](https://github.com/user-attachments/assets/1840a9a8-d0b6-4c49-9c3e-f17694249d5f)
Makes me think of this block, which I added and then later removed from the game last year

**Reply by evanbowman** [reply]
The problem was that getting your crew back was somewhat difficult :)

**Reply by evanbowman** [reply]
An upgraded transporter is a neat idea, it would need to be more powerful in some way than just putting two transporters next to each other, I'll think about it

**Reply by Doronimmo** [reply]
> Makes me think of this block, which I added and then later removed from the game last year

some of the textures are still in-game. (the landing is hilarious)

**Comment by evanbowman** [comment]
I started moving some strings to a separate file, to avoid copy-pasting level scenarios and duplicating code. But I'm having second thoughts about moving the dialog to another file. While convenient, it makes the level scripts difficult to read.

**Reply by Doronimmo** [reply]
Keeping all the dialog in a separate file makes identifying events significantly harder.

**Reply by evanbowman** [reply]
Yeah, I'm going to change it back

**Comment by Doronimmo** [comment]
Quest idea - quest/3_goblin: Some weirdow tells you that there are some goblins on the surface, exposed to extreme radiation, could be a trap.

**Comment by Doronimmo** [comment]
Possible enhancement - unique music tracks for each faction (might no be feasible since sound files take a lot of space)

**Reply by evanbowman** [reply]
It's a good idea! Unfortunately, I'm not sure there's enough space left in the cartridge for this. Maybe in the future if I feel that the game is finished, I'll consider adding more music if there's space

**Comment by evanbowman** [comment]
![raider_waypoint](https://github.com/user-attachments/assets/cafdb742-8b11-4ac1-8b63-5bee20a13113)
Now I'm going back and drawing intro images for the new goblin events and quests

**Reply by Doronimmo** [reply]
![397935229-cafdb742-8b11-4ac1-8b63-5bee20a13113](https://github.com/user-attachments/assets/6d484fbe-e606-43b5-9e7b-b08ee02a6c81)
Is that a smiley face?

**Reply by evanbowman** [reply]
yeah there's some kind of creature in the doorway

**Reply by evanbowman** [reply]
![flares](https://github.com/user-attachments/assets/655cf0c4-983a-4cdc-8848-13f8d5d625f5)
For quest/goblin/3

**Comment by Doronimmo** [comment]
Quest idea - some kind of collector that will reward you for any block in game once (saved between runs). Once given all blocks turns hostile for a mini boss battle (spawns only in zone 3/4). This could be a nice way to add some lore and encourage players to experience more of the game (theoretically only completable with minimum 3 runs (one for each faction))
